File: physics_kapla.py
# Physics kapla towers

import harfang as hg
from math import pi, cos, sin, asin

hg.InputInit()
hg.WindowSystemInit()

res_x, res_y = 1280, 720
win = hg.RenderInit('Harfang - Kapla - Press SPACEBAR', res_x, res_y, hg.RF_VSync | hg.RF_MSAA4X)

hg.AddAssetsFolder('resources_compiled')

pipeline = hg.CreateForwardPipeline()
res = hg.PipelineResources()

# create models
vtx_layout = hg.VertexLayoutPosFloatNormUInt8()

sphere_mdl = hg.CreateSphereModel(vtx_layout, 0.5, 12, 24)
sphere_ref = res.AddModel('sphere', sphere_mdl)

# create materials
prg_ref = hg.LoadPipelineProgramRefFromAssets('core/shader/pbr.hps', res, hg.GetForwardPipelineInfo())

mat_cube = hg.CreateMaterial(prg_ref, 'uBaseOpacityColor', hg.Vec4I(255, 255, 56), 'uOcclusionRoughnessMetalnessColor', hg.Vec4(1, 0.658, 1))
mat_ground = hg.CreateMaterial(prg_ref, 'uBaseOpacityColor', hg.Vec4I(171, 255, 175), 'uOcclusionRoughnessMetalnessColor', hg.Vec4(1, 1, 1))
mat_spheres = hg.CreateMaterial(prg_ref, 'uBaseOpacityColor', hg.Vec4I(255, 71, 75), 'uOcclusionRoughnessMetalnessColor', hg.Vec4(1, 0.5, 0.1))

# setup scene
scene = hg.Scene()
scene.canvas.color = hg.ColorI(200, 210, 208)
scene.environment.ambient = hg.Color.Black

cam = hg.CreateCamera(scene, hg.Mat4.Identity, 0.01, 1000)
scene.SetCurrentCamera(cam)

lgt = hg.CreateLinearLight(scene, hg.TransformationMat4(hg.Vec3(0, 0, 0), hg.Deg3(19, 59, 0)), hg.Color(1.5, 0.9, 1.2, 1), hg.Color(1.5, 0.9, 1.2, 1), 10, hg.LST_Map, 0.002, hg.Vec4(8, 20, 40, 120))
back_lgt = hg.CreatePointLight(scene, hg.TranslationMat4(hg.Vec3(30, 20, 25)), 100, hg.Color(0.8, 0.5, 0.4, 1), hg.Color(0.8, 0.5, 0.4, 1), 0)

mdl_ref = res.AddModel('ground', hg.CreateCubeModel(vtx_layout, 200, 0.1, 200))
hg.CreatePhysicCube(scene, hg.Vec3(200, 0.1, 200), hg.TranslationMat4(hg.Vec3(0, -0.5, 0)), mdl_ref, [mat_ground], 0)


def add_kapla_tower(scn, resources, width, height, length, radius, material, level_count, x, y, z):
	"""Create a Kapla tower, return a list of created nodes"""
	level_y = y + height / 2

	kapla_mdl = hg.CreateCubeModel(vtx_layout, width, height, length)
	kapla_ref = resources.AddModel('kapla', kapla_mdl)

	nodes = []

	for i in range(level_count // 2):
		def fill_ring(r, ring_y, size, r_adjust, y_off):
			step = asin((size * 1.01) / 2 / (r - r_adjust)) * 2
			cube_count = (2 * pi) // step
			error = 2 * pi - step * cube_count
			step += error / cube_count  # distribute error

			a = 0
			while a < (2 * pi - error):
				world = hg.TransformationMat4(hg.Vec3(cos(a) * r + x, ring_y, sin(a) * r + z), hg.Vec3(0, -a + y_off, 0))
				node = hg.CreatePhysicCube(scn, hg.Vec3(width, height, length), world, kapla_ref, [material], 0.1)
				nodes.append(node)
				a += step

		fill_ring(radius - length / 2, level_y, width, length / 2, pi / 2)
		level_y += height
		fill_ring(radius - length + width / 2, level_y, length, width / 2, 0)
		fill_ring(radius - width / 2, level_y, length, width / 2, 0)
		level_y += height

	return nodes


add_kapla_tower(scene, res, 0.5, 2, 2, 6, mat_cube, 12, -12, 0, 0)
add_kapla_tower(scene, res, 0.5, 2, 2, 6, mat_cube, 12, 12, 0, 0)

clocks = hg.SceneClocks()

# input devices and fps controller states
keyboard = hg.Keyboard()
mouse = hg.Mouse()

cam_pos = hg.Vec3(28.3, 31.8, 26.9)
cam_rot = hg.Vec3(0.6, -2.38, 0)

# setup physics
physics = hg.SceneBullet3Physics()
physics.SceneCreatePhysicsFromAssets(scene)
physics_step = hg.time_from_sec_f(1 / 60)

# main loop
while not keyboard.Down(hg.K_Escape):
	keyboard.Update()
	mouse.Update()

	dt = hg.TickClock()

	hg.FpsController(keyboard, mouse, cam_pos, cam_rot, 20 if keyboard.Down(hg.K_LShift) else 8, dt)

	cam.GetTransform().SetPos(cam_pos)
	cam.GetTransform().SetRot(cam_rot)

	if keyboard.Pressed(hg.K_Space):
		node = hg.CreatePhysicSphere(scene, 0.5, hg.TranslationMat4(cam_pos), sphere_ref, [mat_spheres], 0.5)
		physics.NodeCreatePhysicsFromAssets(node)
		physics.NodeAddImpulse(node, hg.GetZ(cam.GetTransform().GetWorld()) * 25.0, cam_pos)

	hg.SceneUpdateSystems(scene, clocks, dt, physics, physics_step, 1)
	hg.SubmitSceneToPipeline(0, scene, hg.IntRect(0, 0, res_x, res_y), True, pipeline, res)

	hg.Frame()
	hg.UpdateWindow(win)

hg.DestroyForwardPipeline(pipeline)

hg.RenderShutdown()
hg.DestroyWindow(win)

File: physics_kapla.go
package main

import (
	"math"
	"runtime"
	"runtime/debug"

	hg "github.com/harfang3d/harfang-go/v3"
)

func fillRing(scn *hg.Scene, material *hg.Material, kaplaRef *hg.ModelRef, width float32, height float32, length float32, r float32, ringY float32, size float32, rAdjust float32, yOff float32, x float32, y float32, z float32) {
	step := math.Asin(float64((size*1.01)/2.0/(r-rAdjust))) * 2.0
	cubeCount := math.Floor((2.0 * math.Pi) / step)
	error := 2.0*math.Pi - step*cubeCount
	step += error / cubeCount // distribute error

	for a := 0.0; a < 2.0*math.Pi-error; a += step {
		world := hg.TransformationMat4(hg.NewVec3WithXYZ(float32(math.Cos(a))*r+x, ringY, float32(math.Sin(a))*r+z), hg.NewVec3WithXYZ(0, float32(-a)+yOff, 0))
		hg.CreatePhysicCubeWithSliceOfMaterialsMass(scn, hg.NewVec3WithXYZ(width, height, length), world, kaplaRef, hg.GoSliceOfMaterial{material}, 0.1)
	}
}

func addKaplaTower(scn *hg.Scene, ressources *hg.PipelineResources, width float32, height float32, length float32, radius float32, material *hg.Material, levelCount int, x float32, y float32, z float32, vtxLayout *hg.VertexLayout) {
	// Create a Kapla tower, return a list of created nodes
	levelY := y + height/2.0

	kaplaMdl := hg.CreateCubeModel(vtxLayout, width, height, length)
	kaplaRef := ressources.AddModel("kapla", kaplaMdl)

	for i := 0; i < levelCount/2; i++ {
		fillRing(scn, material, kaplaRef, width, height, length, radius-length/2.0, levelY, width, length/2.0, math.Pi/2.0, x, y, z)
		levelY += height
		fillRing(scn, material, kaplaRef, width, height, length, radius-length+width/2.0, levelY, length, width/2.0, 0, x, y, z)
		fillRing(scn, material, kaplaRef, width, height, length, radius-width/2.0, levelY, length, width/2.0, 0, x, y, z)
		levelY += height
	}
}

func main() {
	debug.SetGCPercent(-1)

	hg.InputInit()
	hg.WindowSystemInit()

	resX, resY := int32(1280), int32(720)

	win := hg.RenderInitWithWindowTitleWidthHeightResetFlags("Harfang - Kapla", resX, resY, hg.RFVSync|hg.RFMSAA4X)

	hg.AddAssetsFolder("resources_compiled")

	pipeline := hg.CreateForwardPipeline()
	res := hg.NewPipelineResources()

	// create models
	vtxLayout := hg.VertexLayoutPosFloatNormUInt8()

	sphereMdl := hg.CreateSphereModel(vtxLayout, 0.5, 12, 24)
	sphereRef := res.AddModel("sphere", sphereMdl)

	// Create material
	prgRef := hg.LoadPipelineProgramRefFromAssets("core/shader/pbr.hps", res, hg.GetForwardPipelineInfo())

	matCube := hg.CreateMaterialWithValueName0Value0ValueName1Value1(prgRef, "uBaseOpacityColor", hg.Vec4I(255, 255, 56), "uOcclusionRoughnessMetalnessColor", hg.NewVec4WithXYZ(1, 0.658, 1))
	matGround := hg.CreateMaterialWithValueName0Value0ValueName1Value1(prgRef, "uBaseOpacityColor", hg.Vec4I(171, 255, 175), "uOcclusionRoughnessMetalnessColor", hg.NewVec4WithXYZ(1, 1, 1))
	matSpheres := hg.CreateMaterialWithValueName0Value0ValueName1Value1(prgRef, "uBaseOpacityColor", hg.Vec4I(255, 71, 75), "uOcclusionRoughnessMetalnessColor", hg.NewVec4WithXYZ(1, 0.5, 0.1))

	// Setup scene
	scene := hg.NewScene()
	scene.GetCanvas().SetColor(hg.ColorI(200, 210, 208))
	scene.GetEnvironment().SetAmbient(hg.ColorGetBlack())

	cam := hg.CreateCamera(scene, hg.Mat4GetIdentity(), 0.01, 1000)
	scene.SetCurrentCamera(cam)

	hg.CreateLinearLightWithDiffuseSpecularPriorityShadowTypeShadowBiasPssmSplit(scene, hg.TransformationMat4(hg.NewVec3WithXYZ(0, 0, 0), hg.Deg3(19, 59, 0)), hg.NewColorWithRGBA(1.5, 0.9, 1.2, 1), hg.NewColorWithRGBA(1.5, 0.9, 1.2, 1), 10, hg.LSTMap, 0.002, hg.NewVec4WithXYZW(8, 20, 40, 120))
	hg.CreatePointLightWithDiffuseSpecularPriority(scene, hg.TranslationMat4(hg.NewVec3WithXYZ(30, 20, 25)), 100, hg.NewColorWithRGBA(0.8, 0.5, 0.4, 1), hg.NewColorWithRGBA(0.8, 0.5, 0.4, 1), 0)

	mdlRef := res.AddModel("ground", hg.CreateCubeModel(vtxLayout, 200, 0.1, 200))
	hg.CreatePhysicCubeWithSliceOfMaterialsMass(scene, hg.NewVec3WithXYZ(200, 0.1, 200), hg.TranslationMat4(hg.NewVec3WithXYZ(0, -0.5, 0)), mdlRef, hg.GoSliceOfMaterial{matGround}, 0)

	addKaplaTower(scene, res, 0.5, 2, 2, 6, matCube, 12, -12, 0, 0, vtxLayout)
	addKaplaTower(scene, res, 0.5, 2, 2, 6, matCube, 12, 12, 0, 0, vtxLayout)

	clocks := hg.NewSceneClocks()

	// Input devices and fps controller states
	keyboard := hg.NewKeyboard()
	mouse := hg.NewMouse()

	camPos := hg.NewVec3WithXYZ(28.3, 31.8, 26.9)
	camRot := hg.NewVec3WithXYZ(0.6, -2.38, 0)

	// Setup physics
	physics := hg.NewSceneBullet3Physics()
	physics.SceneCreatePhysicsFromAssets(scene)
	physicsStep := hg.TimeFromSecF(1.0 / 60.0)

	// Main loop
	for !hg.ReadKeyboard().Key(hg.KEscape) && hg.IsWindowOpen(win) {
		keyboard.Update()
		mouse.Update()

		dt := hg.TickClock()

		speed := float32(8.0)
		if keyboard.Down(hg.KLShift) {
			speed = 20.0
		}
		hg.FpsController(keyboard, mouse, camPos, camRot, speed, dt)

		cam.GetTransform().SetPos(camPos)
		cam.GetTransform().SetRot(camRot)

		if keyboard.Pressed(hg.KSpace) {
			node := hg.CreatePhysicSphereWithSliceOfMaterialsMass(scene, 0.5, hg.TranslationMat4(camPos), sphereRef, hg.GoSliceOfMaterial{matSpheres}, 0.5)
			physics.NodeCreatePhysicsFromAssets(node)
			physics.NodeAddImpulseWithWorldPos(node, hg.GetZWithM(cam.GetTransform().GetWorld()).MulWithK(25), camPos)
		}

		// Display scene
		viewID := uint16(0)
		hg.SceneUpdateSystemsWithPhysicsStepMaxPhysicsStep(scene, clocks, dt, physics, physicsStep, 1)
		hg.SubmitSceneToPipelineWithFovAxisIsHorizontal(&viewID, scene, hg.NewIntRectWithSxSyExEy(0, 0, resX, resY), true, pipeline, res)

		hg.Frame()
		hg.UpdateWindow(win)
		runtime.GC()
	}

	hg.DestroyForwardPipeline(pipeline)
	hg.RenderShutdown()
	hg.DestroyWindow(win)
}
